scripts/build.py

```diff
@@ -684,26 +823,137 @@ def get_recently_changed_filenames(dev_dir, out_dir):
     # the filenames and their extensions,
     # e.g base.240933e7564bd72a4dde42ee23260c5f.html
     # If a file gets edited, a different MD5 hash is generated.
-    file_hashes = get_file_hashes(dev_dir)
     recently_changed_filenames = []
     # Currently, we do not hash Python files and HTML files are always re-built.
     FILE_EXTENSIONS_NOT_TO_TRACK = ('.html', '.py',)
-    print "Comparing file hashes of %s and %s" % (dev_dir, out_dir)
-    for file_name, md5_hash in file_hashes.iteritems():
+    for filename, md5_hash in source_dir_hashes.iteritems():
         # Ignore files with certain extensions.
-        if any(file_name.endswith(p) for p in FILE_EXTENSIONS_NOT_TO_TRACK):
+        if any(filename.endswith(p) for p in FILE_EXTENSIONS_NOT_TO_TRACK):
             continue
         final_filepath = _insert_hash(
-            os.path.join(out_dir, file_name), md5_hash)
+            os.path.join(out_dir, filename), md5_hash)
         if not os.path.isfile(final_filepath):
             # Filename with provided hash cannot be found, this file has been
             # recently changed or created since last build.
-            recently_changed_filenames.append(file_name)
-    print ('The following files will be rebuilt due to recent changes: %s' %
-           recently_changed_filenames)
+            recently_changed_filenames.append(filename)
+    if recently_changed_filenames:
+        print ('The following files will be rebuilt due to recent changes: %s' %
+               recently_changed_filenames)
     return recently_changed_filenames
 
 
+def generate_build_tasks_to_build_directory(dirnames_dict, file_hashes):
+    """This function queues up build tasks to build all files in source
+    directory if there is no existing staging directory. Otherwise, selectively
+    queue up build tasks to build recently changed files.
+
+    Args:
+        dirnames_dict: dict(str, str, str). A dict containing directory paths to
+            the directory containing source files to be built, the staging
+            directory and the final directory containing built files.
+        file_hashes: dict(str, str). Dictionary with filepaths as keys and
+            hashes of file content as values.
+
+    Returns:
+        build_tasks: deque(Thread). A deque that contains all build tasks queued
+            to be processed.
+    """
+    source_dir = dirnames_dict['dev_dir']
+    staging_dir = dirnames_dict['staging_dir']
+    out_dir = dirnames_dict['out_dir']
+    build_tasks = collections.deque()
+    if not os.path.isdir(staging_dir):
+        # If there is no staging dir, perform build process on all files.
+        print 'Creating new %s folder' % staging_dir
+        build_tasks += generate_build_tasks_to_build_all_files_in_directory(
+            source_dir, staging_dir, file_hashes)
+    else:
+        # If staging dir exists, rebuild all HTML and Python files.
+        mandatory_file_extensions = ('.html', '.py',)
```

Comment:
(Optional) It's a bit unclear what "mandatory" means in this context. Perhaps "file_extensions_to_always_rebuild"?

Reply:
Renamed var!